python/src/aica_api/client.py

```diff
@@ -54,7 +59,93 @@ def _endpoint(self, endpoint=''):
         """
         if self._protocol is None:
             self.protocol()
-        return f'{self._address}/{self._protocol}/{endpoint}'
+        return self.__raw_endpoint(f'{self._protocol}/{endpoint}')
+
+    def __raw_endpoint(self, endpoint: str) -> str:
+        return f'{self._address}/{endpoint}'
+
+    def __ensure_token(self) -> None:
+        """Authenticate with the API and store the result in self.__token."""
+        err = ' The function call may fail due to lack of authentication.'
+        has_version, is_compatible = self._check_version(
+            None,
+            '>=4.3.0',
+            err_incompatible=err,
+            err_undefined=err,
+        )
+        if not has_version or not is_compatible:
+            return
+        if self.__token is not None:
+            return
+        res = requests.post(self._endpoint('auth/login'), headers={'Authorization': f'Bearer {self.__api_key}'})
+        res.raise_for_status()
+        self.__token = res.json()['token']
+
+    def _sio_auth(self) -> Optional[str]:
+        # FIXME: doesn't handle token expiration
+        if self.__api_key is not None:
+            self.__ensure_token()
+        return self.__token
+
+    @staticmethod
+    def _safe_uri(uri: str) -> str:
+        """
+        Make a string safe for use in a URI by encoding special characters.
+
+        :param uri: The URI to sanitize
+        :return: The sanitized URI
+        """
+        return urllib.parse.quote_plus(uri)
+
+    def _request(
```

Comment:
All endpoints now use this function to make requests, which in turn checks that the version is >=v4.3.0 in `ensure_token` so this is a breaking change for this client and makes the other deprecation warnings kind of obsolete no? @eeberhard @LouisBrunner